initialization project

forbrock opened the pull request.

forbrock committed:
initialization project

forbrock committed:
initialization project

grafgeest commented on src/main/java/com/jdbc/dao/UserDao.java:
```diff
@@ -0,0 +1,93 @@
+        }
+    }
+
+    public ArrayList<User> getUsers() {
```
Use List instead

grafgeest commented on .gitignore:
```diff
@@ -0,0 +1,6 @@
+.idea/*
+target/*
+*.class
+Users.iml
```
Use *.iml

forbrock replied:
accepted

forbrock committed:
Changed ArrayList to List

grafgeest commented on src/main/java/com/jdbc/App.java:
```diff
@@ -0,0 +1,21 @@
+public class App {
+    public static void main(String[] args) {
+        UserDao userDao = new UserDao();
+        Connection connection = userDao.getConnection();
```
connection should be closed during finding user

forbrock replied:
this feature will be fixed with new class and connection pool

forbrock committed:
Merge branch 'codereview' of https://github.com/forbrock/Users into codereview

forbrock committed:
fixed conflicts

forbrock committed:
a lot of changes, main ones are:
	(added new class DBCPDataSource,
	 added enum class SearchOption,
	 added dependecies in the pom.xml,
	 the rest of the files were also changed)

forbrock committed:
modified README file

forbrock committed:
Created factory, allowing to create a connection pool from the selected configuration files (also added, for main and in-memory db).

forbrock committed:
changed properties file name

forbrock committed:
renamed AppTest to UserDaoTest

forbrock committed:
UserDao refactored, DBCPDataSourceFactory fixed a mistake in text

forbrock committed:
added @BeforeEach methods for tests

forbrock committed:
refactored delete method

forbrock committed:
logging implemented

grafgeest commented on src/main/java/com/jdbc/dao/Dao.java:
```diff
@@ -0,0 +1,14 @@
+public interface Dao {
+    boolean save(User userObject) throws SQLException;
+    User find(SearchOption searchOption, Object criterion) throws SQLException;
+    List<User> findAll() throws SQLException;
```
Use generics List <T>

grafgeest commented on src/main/java/com/jdbc/dao/Dao.java:
```diff
@@ -0,0 +1,14 @@
+import java.util.List;
+
+public interface Dao {
+    boolean save(User userObject) throws SQLException;
```
Use generics save(T obj)

grafgeest commented on src/main/java/com/jdbc/dao/UserDao.java:
```diff
@@ -0,0 +1,155 @@
+    DataSource dataSource;
+
+    public UserDao(DBType type) {
+        DBCPDataSourceFactory dbcpFactory = new DBCPDataSourceFactory();
```
Use static factory method

grafgeest commented on src/main/java/com/jdbc/dao/UserDao.java:
```diff
@@ -0,0 +1,155 @@
+        } catch (SQLException e) {
+            logger.error("Something went wrong", e);
+        } catch (NullPointerException e) {
+            logger.error(e.getMessage(), e);
```
Use Optional to avoid NPE

grafgeest commented on src/main/java/com/jdbc/dao/UserDao.java:
```diff
@@ -0,0 +1,155 @@
+        }
+    }
+
+    public Connection getConnection() {
```
Are you going to duplicate this logic in all DAO?

grafgeest commented on src/main/java/com/jdbc/dao/factory/DBCPDataSourceFactory.java:
```diff
@@ -0,0 +1,14 @@
+public class DBCPDataSourceFactory {
+
+    public DataSource getDBCPObject(DBType type) {
+        if (type == null) throw new IllegalArgumentException("Wrong database type: " + null);
```
Use Optional

forbrock replied:
this check is redundant, and was deleted.

grafgeest commented on src/main/java/com/jdbc/model/User.java:
```diff
@@ -0,0 +1,86 @@
+    private String name;
+    private LocalDate birth;
+    private String email;
+    DateTimeFormatter formatter = DateTimeFormatter.ofPattern("d/MM/yyyy");
```
private?

grafgeest commented on src/main/logs/AppLog.log:
```diff
@@ -0,0 +1,5 @@
+2020-09-17 22:43:29 [main] ERROR com.jdbc.dao.UserDao - null
```
I suppose this file was committed by mistake

forbrock replied:
exactly

grafgeest commented on src/test/java/com/jdbc/UserDaoTest.java:
```diff
@@ -0,0 +1,151 @@
+    private static UserDao userDao = new UserDao(DBType.H2_INMEMORY_DB);
+    DateTimeFormatter formatter = DateTimeFormatter.ofPattern("d/MM/yyyy");
+
+    @BeforeEach
```
@AfterEach?

Besandr commented on README.md:
```diff
@@ -1 +1,12 @@
+# Users
+CRUD implementation
+How to use methods:
+- _save(userObject)_, takes user object and saves it to the database, the id is adds automatically;
```
Readme instructions should be used for describing how to install and use the application (not a Class). You shouldn't confuse javadocs with Readme.

forbrock committed:
control point

forbrock committed:
changed archetecture for connection pool creation

forbrock committed:
changed archetecture for connection pool creation continues

grafgeest commented on src/main/java/com/jdbc/dao/UserDao.java:
```diff
@@ -0,0 +1,154 @@
+
+    private User queryExecuteHelper(String query) {
+        User user = null;
+        try (Connection conn = getConnection();
```
Connection conn = dataSource.getConnection()

forbrock replied:
fixed

grafgeest commented on src/main/java/com/jdbc/dao/UserDao.java:
```diff
@@ -0,0 +1,154 @@
+
+public class UserDao implements Dao<User> {
+    private static final Logger logger = LoggerFactory.getLogger(UserDao.class);
+    DataSource dataSource;
```
private?

grafgeest commented on src/main/java/com/jdbc/App.java:
```diff
@@ -0,0 +1,22 @@
+
+public class App {
+    private static final Logger logger = LoggerFactory.getLogger(App.class);
+    private static DataSource dataSource = DBCPDataSourceFactory.getDataSource(DBType.POSTGRESQL);
```
put it in main method

forbrock committed:
removed redundant getConnection() method from UserDao

forbrock committed:
some refactoring

forbrock committed:
refactored
